src/main/java/org/prebid/server/bidder/sonobi/SonobiBidder.java

```diff
@@ -1,20 +1,120 @@
+                final ExtImpSonobi extImpSonobi = parseImpExt(imp);
+                final Imp modifiedImp = imp.toBuilder().tagid(extImpSonobi.getTagId()).build();
+                requests.add(makeRequest(bidRequest, modifiedImp));
+            } catch (PreBidException e) {
+                errors.add(BidderError.badInput(e.getMessage()));
+            }
+        }
+
+        return Result.of(requests, errors);
+    }
+
+    private ExtImpSonobi parseImpExt(Imp imp) throws PreBidException {
+        try {
+            return mapper.mapper().convertValue(imp.getExt(), SONOBI_EXT_TYPE_REFERENCE).getBidder();
+        } catch (IllegalArgumentException e) {
+            throw new PreBidException(e.getMessage(), e);
+        }
+    }
+
+    private HttpRequest<BidRequest> makeRequest(BidRequest bidRequest, Imp imp) {
+        final BidRequest modifiedBidRequest = bidRequest.toBuilder().imp(Collections.singletonList(imp)).build();
+
+        return HttpRequest.<BidRequest>builder()
+                .method(HttpMethod.POST)
+                .headers(HttpUtil.headers())
+                .uri(endpointUrl)
+                .body(mapper.encodeToBytes(modifiedBidRequest))
+                .payload(modifiedBidRequest)
                 .build();
     }
+
+    @Override
+    public Result<List<BidderBid>> makeBids(HttpCall<BidRequest> httpCall, BidRequest bidRequest) {
+        try {
+            final BidResponse bidResponse = mapper.decodeValue(httpCall.getResponse().getBody(), BidResponse.class);
+            return Result.withValues(extractBids(httpCall.getRequest().getPayload(), bidResponse));
+        } catch (DecodeException e) {
+            return Result.withError(BidderError.badServerResponse(e.getMessage()));
+        }
+    }
+
+    private static List<BidderBid> extractBids(BidRequest bidRequest, BidResponse bidResponse) {
+        if (bidResponse == null || CollectionUtils.isEmpty(bidResponse.getSeatbid())) {
+            return Collections.emptyList();
+        }
+
+        final String currency = bidResponse.getCur();
+        final Map<String, BidType> impIdToBidType = bidRequest.getImp().stream()
+                .collect(Collectors.toMap(Imp::getId, SonobiBidder::resolveBidType));
+
+        return bidResponse.getSeatbid().stream()
+                .filter(Objects::nonNull)
+                .map(SeatBid::getBid)
+                .filter(Objects::nonNull)
+                .flatMap(Collection::stream)
+                .filter(Objects::nonNull)
+                .map(bid -> BidderBid.of(bid, impIdToBidType.getOrDefault(bid.getImpid(), BidType.banner), currency))
+                .collect(Collectors.toList());
+    }
+
+    private static BidType resolveBidType(Imp imp) {
+        return imp.getBanner() != null ? BidType.banner
+                : imp.getVideo() != null ? BidType.video
+                : imp.getXNative() != null ? BidType.xNative
+                : imp.getAudio() != null ? BidType.audio
+                : BidType.banner;
+    }
```

Comment:
From .yaml file we can see that this bidder uses only banner and video. So let's implement resolveBidType with similar logic to Go implementation 
```java
 private static BidType resolveBidType(String impId, List<Imp> imps) {
        for (Imp imp : imps) {
            if (imp.getId().equals(impId)) {
                if (imp.getBanner() == null && imp.getVideo() != null) {
                    return BidType.video;
                }
                return BidType.banner;
            }
        }
        throw new PreBidException(String.format("Failed to find impression for ID: %s", impId));
    }
```